not related - problems with IAR licence,

```
Compile [ 20.0%]: ip_fsc.c
Compile [ 20.1%]: common_functions.c
[ERROR] Fatal error[LMS001]: License check failed. Use the IAR License Manager to
```

```
Compile [ 12.5%]: QUECTEL_BG96_CellularStack.cpp
[ERROR] 
Failed to build library
```

--- a/events/UserAllocatedEvent.h
+++ b/events/UserAllocatedEvent.h
@@ -243,7 +243,7 @@ public:
      */
     bool cancel()
     {
-        return equeue_cancel_user_allocated(_equeue, &_e);
+        return _post_ref > 0 ? equeue_cancel_user_allocated(_equeue, &_e) : false;
     }
 
 

--- a/events/UserAllocatedEvent.h
+++ b/events/UserAllocatedEvent.h
@@ -86,7 +86,7 @@ public:
      *  @param f        Function to execute when the event is dispatched
      *  @param args     Arguments to bind to the callback
      */
-    constexpr UserAllocatedEvent(F f,  ArgTs... args) : _e(get_default_equeue_event()), _c(f, args...), _equeue(), _post_ref()
+    constexpr UserAllocatedEvent(F f,  ArgTs... args) : _e(get_default_equeue_event()), _c(f, args...), _delay(), _period(-1), _equeue(), _post_ref()
     {
     }
 
@@ -100,7 +100,7 @@ public:
      *  @param f        Function to execute when the event is dispatched
      *  @param args     Arguments to bind to the callback
      */
-    constexpr UserAllocatedEvent(EventQueue *queue, F f,  ArgTs... args) : _e(get_default_equeue_event()), _c(f, args...), _equeue(&queue->_equeue), _post_ref()
+    constexpr UserAllocatedEvent(EventQueue *queue, F f,  ArgTs... args) : _e(get_default_equeue_event()), _c(f, args...), _delay(), _period(-1), _equeue(&queue->_equeue), _post_ref()
     {
     }
 
@@ -215,7 +215,7 @@ public:
     void delay(int delay)
     {
         MBED_ASSERT(!_post_ref);
-        equeue_event_delay(&_e + 1, delay);
+        _delay = delay;
     }
 
     /** Configure the period of an event
@@ -225,7 +225,7 @@ public:
     void period(int period)
     {
         MBED_ASSERT(!_post_ref);
-        equeue_event_period(&_e + 1, period);
+        _period = period;
     }
 
     /** Cancels posted event
@@ -251,6 +251,8 @@ private:
     friend class EventQueue;
     struct equeue_event _e;
     C _c;
+    int _delay;
+    int _period;
     struct equeue *_equeue;
     uint8_t _post_ref;
 
@@ -260,17 +262,22 @@ private:
             return false;
         }
         core_util_atomic_incr_u8(&_post_ref, 1);
+        equeue_event_delay(&_e + 1, _delay);
+        equeue_event_period(&_e + 1, _period);
         equeue_post_user_allocated(_equeue, &EventQueue::function_call<C>, &_e);
         return true;
     }
 
     bool post_on(EventQueue *queue)
     {
+        MBED_ASSERT(queue);
         if (_post_ref) {
             return false;
         }
         _equeue = &(queue->_equeue);
         core_util_atomic_incr_u8(&_post_ref, 1);
+        equeue_event_delay(&_e + 1, _delay);
+        equeue_event_period(&_e + 1, _period);
         equeue_post_user_allocated(_equeue, &EventQueue::function_call<C>, &_e);
         return true;
     }

--- a/events/source/equeue.c
+++ b/events/source/equeue.c
@@ -24,6 +24,12 @@
 // check if the event is allocaded by user - event address is outside queues internal buffer address range
 #define EQUEUE_IS_USER_ALLOCATED_EVENT(e) ((q->buffer == NULL) || ((uintptr_t)(e) < (uintptr_t)q->buffer) || ((uintptr_t)(e) > ((uintptr_t)q->slab.data)))
 
+// for user allocated events use event id to track event state
+enum {
+    EQUEUE_USER_ALLOCATED_EVENT_STATE_INPROGRESS = 1,
+    EQUEUE_USER_ALLOCATED_EVENT_STATE_DONE = 0          // event canceled or dispatching done
+};
+
 // calculate the relative-difference between absolute times while
 // correctly handling overflow conditions
 static inline int equeue_tickdiff(unsigned a, unsigned b)
@@ -229,7 +235,9 @@ void equeue_dealloc(equeue_t *q, void *p)
         e->dtor(e + 1);
     }
 
-    if (!EQUEUE_IS_USER_ALLOCATED_EVENT(e)) {
+    if (EQUEUE_IS_USER_ALLOCATED_EVENT(e)) {
+        e->id = EQUEUE_USER_ALLOCATED_EVENT_STATE_DONE;
+    } else {
         equeue_mem_dealloc(q, e);
     }
 }
@@ -402,6 +410,7 @@ void equeue_post_user_allocated(equeue_t *q, void (*cb)(void *), void *p)
     unsigned tick = equeue_tick();
     e->cb = cb;
     e->target = tick + e->target;
+    e->id = EQUEUE_USER_ALLOCATED_EVENT_STATE_INPROGRESS;
 
     equeue_enqueue(q, e, tick);
     equeue_sema_signal(&q->eventsema);
@@ -424,7 +433,7 @@ bool equeue_cancel(equeue_t *q, int id)
 
 bool equeue_cancel_user_allocated(equeue_t *q, void *e)
 {
-    if (!e) {
+    if (!e || ((struct equeue_event *)e)->id == EQUEUE_USER_ALLOCATED_EVENT_STATE_DONE) {
         return false;
     }
 

--- a/events/source/equeue.c
+++ b/events/source/equeue.c
@@ -506,7 +506,9 @@ void equeue_dispatch(equeue_t *q, int ms)
                 e->target += e->period;
                 equeue_enqueue(q, e, equeue_tick());
             } else {
-                equeue_incid(q, e);
+                if (!EQUEUE_IS_USER_ALLOCATED_EVENT(e)) {
+                    equeue_incid(q, e);
+                }
                 equeue_dealloc(q, e + 1);
             }
         }

